### ISSUE TITLE
packages that use modules with classes don't work unless Rcpp is attached

### ISSUE THREAD

**Issue by s-u** [RcppCore/Rcpp#168]
Packages that use Rcpp modules which expose C++ classes work only if Rcpp is explicitly present on the search path. This has been described, but mis-identified in #118.

To replicate, create a skeleton package via `Rcpp::Rcpp.package.skeleton(module=TRUE)`, build, install and run:

```
> library(anRpackage)
Warning message:
 class "C++Object" is defined (with package slot ‘Rcpp’) but no metadata object found to revise subclass information---not exported?  Making a copy in package ‘anRpackage’ 
> World$new()
Error in .Object$initialize(...) : 
  could not find function "cpp_object_initializer"
```

Note that the skeleton uses `Imports: Rcpp` which is crucial here. If `Rcpp` is explicitly attached, it works:

```
> library(Rcpp)
> library(anRpackage)
> World$new()
C++ object <0x1136d40> of class 'World' <0xa6c8b0>
```

This turns out to be a major problem for the following reason: A package can try to work around this problem by using `Depends: Rcpp`, but that only applies if the package is also attached itself. This means that the "I don't work if not attached" property becomes contagious in that it bubbles through the package dependency chain - if a package has ever even one ancestor in the dependency chain that is a package with Rcpp module then it itself becomes unusable unless Rcpp attached - so in theory all packages must depend on Rcpp - whether they use it or not. Even if desirable ;), it's not practical, unfortunately.

**Comment by ha0ye** [comment]
FWIW, I resolved the warning in my own package by making documentation for each of the S4 classes associated with each Rcpp module. A bit of a hack, but it removes the need to attach Rcpp.

**Comment by eddelbuettel** [comment]
Ok,  For reference, do you have a link or example?

**Comment by ha0ye** [comment]
Yep, see https://github.com/ha0ye/rEDM

**Comment by ivan-krukov** [comment]
This issue can be resolved the following to `NAMESPACE` of a package:
```
import(Rcpp)
```

While this does not deal with the issue, it resolved the symptom.
Note: this is needed even if `importFrom(Rcpp)` is present.

**Comment by eddelbuettel** [comment]
We have always recommend the following form:

```r
importFrom(Rcpp, evalCpp)
```

and of course other people did variants around that for no good reason.  _Any_ symbol from Rcpp, once forced to instantiate, should do.    A plain `import(Rcpp)` sometimes works, what we have here works more often and is not really anymore complicated.

**Comment by ivan-krukov** [comment]
Yes, I have taken the snippet from `Rcpp-modules` vignette. I stumbled on this by googling the error. Figured it would be a good place to leave a note.

**Comment by eddelbuettel** [comment]
Maybe that is a good item to add to the Rcpp-FAQ too.

**Comment by ivan-krukov** [comment]
I misunderstood what `importFrom()` actually meant. Updated comment above.

**Comment by eddelbuettel** [comment]
If we knew how/why/where that is needed....   I have packages with modules, and I only do `importFrom(Rcpp, evalCpp)`.  (And I updated mine too now.)

**Comment by ivan-krukov** [comment]
Here is a small example where I see the issue: https://github.com/ivan-krukov/mre168

**Comment by ha0ye** [comment]
Seems related to using modules to instantiate C++ class objects?

**Comment by eddelbuettel** [comment]
Really good minimally reproducible example, thanks for setting it up. 

And I concur: it seems one needs to import Rcpp.

**Comment by eddelbuettel** [comment]
I have one [counterexample in RcppXts](https://github.com/eddelbuettel/rcppxts/blob/master/NAMESPACE#L1-L5) -- but you could argue that its use of Modules is lighter-weight and does not instantiate classes (just how @ha0ye surmised).

**Comment by kevinushey** [comment]
Some debugging information about the environment that `initialize` uses. In the working case:

```
Browse[6]> pryr::parenvs(environment())
  label                           name            
1 <environment: 0x103657db8>      ""              
2 <environment: 0x106cac678>      ""              
3 <environment: namespace:mre168> ""              
4 <environment: 0x101d51910>      "imports:mre168"
5 <environment: namespace:base>   ""              
6 <environment: R_GlobalEnv>      ""     

Browse[6]> pryr::where("cpp_object_initializer")
<environment: 0x101d51910>
attr(,"name")
[1] "imports:mre168"
```

In the 'not working' case, it looks like the `mre168` namespace doesn't end up importing the required symbol `cpp_object_initializer`. I'm not sure what `methods` machinery comes into play to ensure that the namespace imports library is appropriately populated (or if Rcpp is doing its own hackery around this)

**Comment by kevinushey** [comment]
tl;dr: packages that want to use modules must import the entire Rcpp namespace, since the functions invoked will depend on objects on that namespace (and so we need to make sure that they're imported by the client package)

**Comment by eddelbuettel** [comment]
Yeppers.  We probably need to straighten the documentation out.  Strictly speaking it is _most_ packages using Modules (see my RcppXts counterexample), but we may as well recommended it outright.

**Comment by coatless** [comment]
@eddelbuettel: Doc flag please + 0.12.9 label?

**Comment by kevinushey** [comment]
Whoops, looks like I can set milestones but not add new ones.

As an aside, we might also want to update the behavior of `Rcpp.package.skeleton()` to ensure the NAMESPACE file is populated as such when modules are used.

**Comment by eddelbuettel** [comment]
We have this in [Rcpp.package.skeleton](https://github.com/RcppCore/Rcpp/blob/master/R/Rcpp.package.skeleton.R) 

```r
    if (isTRUE(module)) {
        writeLines('import(methods, Rcpp)', ns)
        message(" >> added import(methods, Rcpp) directive to NAMESPACE")
    } else {
        writeLines('importFrom(Rcpp, evalCpp)', ns)
        message(" >> added importFrom(Rcpp, evalCpp) directive to NAMESPACE" )
    }
```

and the `import(methods, Rcpp)` is actually what most of my packages using Modules.  Maybe no need for change after all.   We _already_ special-case Modules.

**Comment by ha0ye** [comment]
3.1.1 of the modules vignette already recommends "import(Rcpp)" for the NAMESPACE of packages that use Rcpp modules.

**Comment by kevinushey** [comment]
Well then! Perhaps the only thing that may be warranted here is an FAQ entry?

**Comment by eddelbuettel** [comment]
... saying "just trust `Rcpp.package.skeleton(pkgName, module=TRUE)` ?

Don't know.  Maybe we are actually done. Code seems correct, documentation seems correct.

**Comment by ha0ye** [comment]
I followed the vignette and didn't have a problem. At one point I tried "importFrom(Rcpp, evalCpp)", found it didn't work, and reverted. Curious to know how @ivan-krukov stumbled onto the issue to see if there's a gap somewhere that should be patched.

**Comment by ivan-krukov** [comment]
@ha0ye My example was mostly taken from the auto-generated RStudio "Package with Rcpp" template.

**Comment by eddelbuettel** [comment]
Yes, that is more between you and RStudio for thinking it would just work with Modules.  But it needs subtle  adjustments.  

I'm in favor of closing this.

**Comment by kevinushey** [comment]
+1 (I'll double check to see if we need to change anything on the RStudio side)

**Comment by eddelbuettel** [comment]
If you have 'infinite capacity' another toggle for 'Package w/Rcpp'  _and modules_ could help.  
But there are too many options as it stands.

**Comment by eddelbuettel** [comment]
Closing this now though. This seems known, documented and supported in the code (even though some of us, yours truly included, occasionally give less perfect recommendations).
